app/science_module/src/main.cpp

```diff
@@ -0,0 +1,189 @@
+
+static DigitalOut* digital_out_vector[NUM_MODULES] = {elevator_dir_pin,
+                                                      drill_dir_pin,
+                                                      temp_dir_pin,
+                                                      flap_dir_pin};
+
+/*
+ *  Controls for each instrument.
+ *  Positive float is high and negative float is low.
+ */
+void control_modules(float pwm_duty, uint16_t module_id)
+{
+    pwm_out_vector[module_id]->write(fabs(pwm_duty));
+    *digital_out_vector[module_id] = (pwm_duty >= 0);
+}
+
+const char am2315_temp_humidity_sensor = 0xB8;
+
+/*
+ * Get temp/humidity sensor readings from i2c
+ */
+uint32_t get_temp_humidity_sensor_reading()
+{
+    char sensor_cmd[4] = {0};
+    uint32_t *sensor_data;
+    
+    // Host to sensor read temp and humidity data
+    sensor_cmd[0] = 0x03;
+    sensor_cmd[1] = 0x00;
+    sensor_cmd[2] = 0x04;
+    i2c.write(am2315_temp_humidity_sensor, sensor_cmd, 3);
+
+    // Tell sensor to prep sensor data
+    sensor_cmd[0] = 0x03;
+    sensor_cmd[1] = 0x04;
+    i2c.write(am2315_temp_humidity_sensor, sensor_cmd, 2);
+
+    wait(0.5);
+
+    // Read data from sensor
+    memset(sensor_cmd, 0, sizeof(sensor_cmd)/sizeof(sensor_cmd[0]));
+    i2c.read(am2315_temp_humidity_sensor, sensor_cmd, 4);
+
+    sensor_data = (uint32_t*) &sensor_cmd[0];
+    return *sensor_data;
+}
+
+/*
+ *  Receives pwm data from CAN bus
+ */
+void CANLIB_Rx_OnMessageReceived(void)
+{
+    can_id = CANLIB_Rx_GetSenderID();
+    switch(can_id){
+        case elevator_pwm_filter_ID:
+        case drill_pwm_filter_ID:
+        case temp_pwm_filter_ID:
+        case flap_pwm_filter_ID:
+            pwm_duty = CANLIB_Rx_GetAsFloat(CANLIB_INDEX_0);
+            can_write_pwm = true;
+            break;
+        case temp_humidity_filter_ID:
+            can_read_temp_sensor = true;
+            break;
+    }
+}
+
+module_ids_t init_modules[INIT_VECTOR_SIZE] = {elevator_pwm_filter_ID,
+                                                drill_pwm_filter_ID,
+                                                temp_pwm_filter_ID,
+                                                flap_pwm_filter_ID,
+                                                temp_humidity_filter_ID};
+
+int main()
+{
+    uint32_t temp_humidity_sensor_reading;
+    float temp_reading;
+    float humidity_reading;
+    /*
+     * Initialize CAN for pwm
+     */
+    if (CANLIB_Init(elevator_pwm_ID, CANLIB_LOOPBACK_OFF) != 0)
+    {
+        pc.printf("CAN Initialization Failed\r\n");
+    }
+
+    for(int i = 0; i < INIT_VECTOR_SIZE; i++)
+    {
+        if (CANLIB_AddFilter(init_modules[i]) != 0)
+        {
+            pc.printf("CAN Add Filter Failed on Filter: %u \r\n", init_modules[i]);
+        }
+    }
+
+    /*
+     * Loop
+     */
+    while(1)
+    {
+        if(can_write_pwm)
+        {
+            control_modules(pwm_duty, can_id % science_module_pwm_filter);
```

Comment:
don't think this math will work because `can_id` will be 420-423 and your `science_module_pwm_filter` is 400. this will cause you to send 20-23 which are invalid indexes